src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1461 @@
+        # [ngram, B*head, T, 2*T]
+        attn_weights_ngram = attn_weights_ngram + predict_relative_logits
+
+        if ngram_mask_matrix is not None:
+            ngram_mask_matrix = ngram_mask_matrix.unsqueeze(1)
+            attn_weights_ngram = attn_weights_ngram + ngram_mask_matrix
+
+        attn_weights_ngram = softmax(
+            attn_weights_ngram, dim=-1, onnx_trace=self.onnx_trace,
+        ).type_as(attn_weights_ngram)
+        attn_weights_ngram = F.dropout(attn_weights_ngram, p=self.dropout, training=self.training)
+
+        # [ngram, B*head, T, c]
+        attn_ngram = torch.einsum('nbts,nbsc->nbtc', (attn_weights_ngram, v_ngram))
+        # [ngram, T, B, C]
+        attn_ngram = attn_ngram.transpose(1, 2).contiguous().view(self.ngram, real_tgt_len, bsz, embed_dim)
+        attn_ngram = self.out_proj(attn_ngram)
+
+        attn_result = []
+        attn_result.append(attn_main)
+        attn_result.append(attn_ngram)
+
+        # [1+ngram*T, B, C]
+        attn = torch.cat(attn_result, 0).view(-1, bsz, embed_dim)
+
+        if output_attentions:
+            attn_weights = attn_weights_ngram#.view(bsz, self.num_heads, tgt_len, src_len)
+        else:
+            attn_weights = None
+
+        return attn, attn_weights
+
+    def in_proj_qkv(self, query):
+        return self._in_proj(query).chunk(3, dim=-1)
+
+    def in_proj_q(self, query):
+        if self.qkv_same_dim:
+            return self._in_proj(query, end=self.embed_dim)
+        else:
+            bias = self.in_proj_bias
+            if bias is not None:
+                bias = bias[:self.embed_dim]
+            return F.linear(query, self.q_proj_weight, bias)
+
+    def in_proj_k(self, key):
+        if self.qkv_same_dim:
+            return self._in_proj(key, start=self.embed_dim, end=2 * self.embed_dim)
+        else:
+            weight = self.k_proj_weight
+            bias = self.in_proj_bias
+            if bias is not None:
+                bias = bias[self.embed_dim:2 * self.embed_dim]
+            return F.linear(key, weight, bias)
+
+    def in_proj_v(self, value):
+        if self.qkv_same_dim:
+            return self._in_proj(value, start=2 * self.embed_dim)
+        else:
+            weight = self.v_proj_weight
+            bias = self.in_proj_bias
+            if bias is not None:
+                bias = bias[2 * self.embed_dim:]
+            return F.linear(value, weight, bias)
+
+    def _in_proj(self, input, start=0, end=None):
+        weight = self.in_proj_weight
+        bias = self.in_proj_bias
+        weight = weight[start:end, :]
+        if bias is not None:
+            bias = bias[start:end]
+        return F.linear(input, weight, bias)
+
+    def reorder_incremental_state(self, incremental_state, new_order):
+        """Reorder buffered internal state (for incremental generation)."""
+        input_buffer = self._get_input_buffer(incremental_state)
```

Comment:
`input_buffer = {}` instead or is `_get_input_buffer` not fully implemented yet?